gokart/redis_lock.py

```diff
@@ -0,0 +1,77 @@
+import os
+from logging import getLogger
+from typing import NamedTuple, Optional
+
+import redis
+from apscheduler.schedulers.background import BackgroundScheduler
+
+logger = getLogger(__name__)
+
+
+class RedisParams(NamedTuple):
+    redis_host: str = None
+    redis_port: str = None
+    redis_key: str = None
+    should_redis_lock: bool = False
+
+
+class RedisClient:
+    _instances: dict = {}
+
+    def __new__(cls, *args, **kwargs):
+        key = (args, tuple(sorted(kwargs.items())))
+        if cls not in cls._instances:
+            cls._instances[cls] = {}
+        if key not in cls._instances[cls]:
+            cls._instances[cls][key] = super(RedisClient, cls).__new__(cls)
+        return cls._instances[cls][key]
+
+    def __init__(self, host: str, port: str) -> None:
+        if not hasattr(self, '_redis_client'):
+            self._redis_client = redis.Redis(host=host, port=port)
+
+    def get_redis_client(self):
+        return self._redis_client
+
+
+def with_lock(func, redis_params: RedisParams):
+    if not redis_params.should_redis_lock:
+        return func
+
+    def wrapper(*args, **kwargs):
+        redis_client = RedisClient(host=redis_params.redis_host, port=redis_params.redis_port).get_redis_client()
+        redis_lock = redis.lock.Lock(redis=redis_client, name=redis_params.redis_key, timeout=30, blocking=True, thread_local=False)
```

Comment:
I think `timeout` is important parameter.
It's better to be able to set it up externally.